### ISSUE TITLE
EPIC: Permissioned Namespaces

### ISSUE THREAD

**Issue by evan-forbes** [celestiaorg/celestia-app#3682]
Per sync convos at the modsummit, it was suggested that one tool against the woods attack for usecases that have an set of allowed blob posters would be to provide a new namespace (or perhaps share version?) would be to have a namespace for each account where only that account could post to that namespace. Notably, this would include multisigs, so 1/7 multisig could be created would allow only for any of the 7 multisig members to post to that namespace. 

One flavor of this expands the scope of permissioned namesapces to not be limited by a multisig, instead having an arbitrary number of allowed posters. 

s/o to @hdevalence for the idea! The suggested usecase afaiu was for privacy focused threshold signing such as [FROST](https://www.tftc.io/issue-1379-using-frost-to-increase-privacy-in-collaborative-bitcoin-custody-models/). Here users could be post shards to the namespace and all participants would have timestamped and published copies of all shards

- [ ] Write CIP
- [ ] TODO: update list of issues following the CIP

**Comment by hdevalence** [comment]
I'm happy to help write a CIP — I was going to make a forum post today before seeing this issue — happy to DM about drafting

**Comment by cmwaters** [comment]
Yeah I think the simplest approach would be to use account numbers for uniqueness and a separate namespace version. If multisigs work and aren't a pain in the ass to use then this could also be sufficient for supporting groups. The other idea may be to use the `x/group` module to have multiply accounts represented as a single account, which would also allow for the group set to change

**Comment by S1nus** [comment]
Instead of adding logic to restrict namespaces, could we:
1. sort the blobs numerically by signer
2. encode min-max signer in the NMT, in addition to min-max namespace
3. filter by signer + namespace, the same way we can filter by namespace

**Comment by smsunarto** [comment]
I'm in support of this too. Specifically, I would prefer for there to be a direct permissioned namespace in contrast to signer-partitioned namespace.

The most important part for me here is more from the perspective of developer (and potentially user experience).

1. Arbitrary signers being able to add blob to the namespace adds a dangerous footgun where a developer might assume that the namespace is exclusively owned by him and all blobs can be trusted, which is not the case.

2. Blob explorer like Celenium labels rollup based on their namespace for easy discovery, but if there's multiple rollups posting blobs to the namespace (let it be intentionally or unintentionally) then this will create confusion. It will also add additional complexity for such explorer to parse out which blob belongs to which rollup. Signer partitioned (and versions) will make this worse imo.

3. Partitioning blobs based on signers will make doing key rotations more difficult since now it would be necessary for rollups to hardcode the previous signer's info to fetch the correct blobs.

Happy to collaborate on the CIP too since this is on the top of our bucket list.

**Comment by hdevalence** [comment]
> Instead of adding logic to restrict namespaces, could we:
> 
>     1. sort the blobs numerically by signer
> 
>     2. encode min-max signer in the NMT, in addition to min-max namespace
> 
>     3. filter by signer + namespace, the same way we can filter by namespace

No, because the goal is being able to permission the namespace. 

This still allows the namespace to be filled with garbage, but it forces every single downstream consumer to be responsible for filtering it.

This is both more technically complicated to build and also worse from a dev UX standpoint.

To be honest, I'm a little surprised that this isn't already possible; I can't start building the rollups I'd like to build until it lands, so my preference would be to ship the simplest possible thing as fast as possible.

**Comment by musalbas** [comment]
This is also related (and can be thought as a specific version) of conditional namespaces: https://forum.celestia.org/t/customize-your-namespace-rollup-resource-pricing-account-abstraction-and-native-restaking/1666
cc @nashqueue 

In terms of longer term "garbage attack" mitigations, ideally we need [intra-namespace blob prioritization](https://github.com/celestiaorg/celestia-app/issues/1274), i.e. blobs are ordered by gas paid in the namespace, so that indexers can ignore "spam" blobs below a certain fee.

If we want to allow censorship-resistant multi-user/sequencer rollups, I'm not sure there's any way to avoid requiring indexers to parse potential spam. Consider the fact that in an forced transaction inclusion model for an Ethereum rollup for example, users can spam that inbox with low fee transactions.

That being said, I'm not inherently against permissioned single-user namespaces if there are good use cases for them (such as no need for single-user censorship-resistance in @hdevalence's use case), but they should not be used as an excuse for other use cases to enshrine a single permanent sequencer by default for multi-user rollups, as that wouldn't be censorship-resistant for those rollups...

**Comment by hdevalence** [comment]
The concern about permissioning being used as an excuse to enshrine a sequencer is a good one, but conversely, it would also be suboptimal if the concern about enshrinement prevented a simple solution to the problem from shipping now, and enabling people build on Celestia.

**Comment by musalbas** [comment]
Yeah I'm not against it if there's user demand for it, but I wanted to caveat it with the above

**Comment by S1nus** [comment]
> forces every single downstream consumer to be responsible for filtering it

FullNodes already filter the square when you call `GetAll` on a namespace. But yea i agree, it would be harder to change the tree. Just throwing the idea out there in case we want to avoid adding something to the state transition function.
